sklearn/metrics/classification.py

```diff
@@ -391,72 +391,114 @@ def jaccard_similarity_score(y_true, y_pred, normalize=True,
-        coefficient over the sample set.
-
-        The best performance is 1 with ``normalize == True`` and the number
-        of samples with ``normalize == False``.
+    score: float (if average is not None) or array of float, shape =\
+            [n_unique_labels]
 
     See also
     --------
     accuracy_score, hamming_loss, zero_one_loss
 
     Notes
     -----
-    In binary and multiclass classification, this function is equivalent
-    to the ``accuracy_score``. It differs in the multilabel classification
-    problem.
+    In differs in implementation from ``accuracy_score`` from all three
+    classifications i.e. binary, mutliclass and multilabel.
 
     References
     ----------
     .. [1] `Wikipedia entry for the Jaccard index
            <https://en.wikipedia.org/wiki/Jaccard_index>`_
 
-
     Examples
     --------
     >>> import numpy as np
     >>> from sklearn.metrics import jaccard_similarity_score
     >>> y_pred = [0, 2, 1, 3]
     >>> y_true = [0, 1, 2, 3]
-    >>> jaccard_similarity_score(y_true, y_pred)
+    >>> jaccard_similarity_score(y_true, y_pred, average='macro')
     0.5
-    >>> jaccard_similarity_score(y_true, y_pred, normalize=False)
-    2
+    >>> jaccard_similarity_score(y_true, y_pred, normalize='micro')
+    0.33...
 
     In the multilabel case with binary label indicators:
 
     >>> jaccard_similarity_score(np.array([[0, 1], [1, 1]]),\
         np.ones((2, 2)))
     0.75
+
+    In the multiclass case:
+
+    >>> y_pred = ['ant', 'ant', 'cat', 'cat', 'ant', 'cat']
+    >>> y_true = ['cat', 'ant', 'cat', 'cat', 'ant', 'bird']
+    >>> jaccard_similarity_score(y_true, y_pred)
+    0.38888888888888884
     """
 
+    average_options = (None, 'micro', 'macro', 'weighted')
+    if average not in average_options:
+        raise ValueError("average has to be one of " + str(average_options))
+
     # Compute accuracy for each possible representation
     y_type, y_true, y_pred = _check_targets(y_true, y_pred)
     check_consistent_length(y_true, y_pred, sample_weight)
+
     if y_type.startswith('multilabel'):
         with np.errstate(divide='ignore', invalid='ignore'):
-            # oddly, we may get an "invalid" rather than a "divide" error here
-            pred_or_true = count_nonzero(y_true + y_pred, axis=1)
-            pred_and_true = count_nonzero(y_true.multiply(y_pred), axis=1)
-            score = pred_and_true / pred_or_true
-            score[pred_or_true == 0.0] = 1.0
+            pred_or_true = count_nonzero(y_true + y_pred, axis=0)
+            pred_and_true = count_nonzero(y_true.multiply(y_pred), axis=0)
+            if average == 'macro':
+                score = pred_and_true / pred_or_true
+                n_features = y_true.shape[1]
+                return np.sum(score) / n_features
+            elif average == 'micro':
+                score = np.sum(pred_and_true) / np.sum(pred_or_true)
+                return score
+            elif average == 'weighted':
```

Comment:
How would the meaning of 'weighted' fit in this?

From what I understand 'average' set to any value should use `axis=0` sum not `axis=1` since the definition of 'macro', 'micro', ... involve "calculate metric for each label" (i.e. for each category), right?

As you can see:
```
'macro':

    Calculate metrics for each label, and find their unweighted mean. This does not take label imbalance into account.
```